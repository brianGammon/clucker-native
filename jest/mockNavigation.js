/* @flow */
type ReturnVals = {
  [methodName: string]: {
    [param: string]: string,
  },
};

type MockNavigation = {
  getParam: (param: string, defaultId: string) => string,
};

const mockNavigation = (returnVals: ReturnVals) => ({
  getParam: jest.fn((param: string, defaultId: string) => {
    if (returnVals.getParam && returnVals.getParam[param]) {
      return returnVals.getParam[param];
    }
    return defaultId;
  }),
}: MockNavigation);

export default mockNavigation;
