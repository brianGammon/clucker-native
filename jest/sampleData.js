const sampleData1 = {
  userSettings: {
    key: 'user1',
    data: {
      currentFlockId: 'flock1',
    },
  },
  flocks: {
    data: {
      flock1: {
        name: 'Test Flock 1',
        ownedBy: 'user1',
      },
      flock2: {
        name: 'Test Flock 2',
        ownedBy: 'user2',
      },
    },
  },
  chickens: {
    data: {
      chicken1: {
        name: 'The First Chicken',
      },
      chicken2: {
        name: 'The Second Chicken',
      },
    },
  },
  eggs: {
    data: {
      egg1: {
        chickenId: 'chicken1',
        chickenName: 'Bossie',
        damaged: false,
        date: '2018-08-01',
        modified: '2018-01-12T18:05:54.806Z',
        notes:
          'This was the record setting egg from May 2017, real egg was 66.1g on this day',
        userId: 'MVQUEcoTaVUhUdo2ZS0zsIE6tn93',
        weight: 88.1,
      },
      egg2: {
        chickenId: 'chicken2',
        chickenName: 'Baby',
        damaged: false,
        date: '2017-09-24',
        modified: '2017-10-12T18:05:54.806Z',
        userId: 'MVQUEcoTaVUhUdo2ZS0zsIE6tn93',
        weight: 88.1,
      },
      egg3: {
        chickenId: 'chicken1',
        chickenName: 'Bossie',
        damaged: false,
        date: '2018-08-02',
        modified: '2018-08-02T18:05:54.806Z',
        notes: '',
        userId: 'MVQUEcoTaVUhUdo2ZS0zsIE6tn93',
        weight: 68.1,
      },
    },
  },
};

export default sampleData1;
