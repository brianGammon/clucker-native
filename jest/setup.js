/* eslint-disable import/no-extraneous-dependencies */
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

global.shallow = shallow;
global.mount = mount;

jest.mock('Linking', () => ({
  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
  openURL: jest.fn(),
  canOpenURL: jest.fn(),
  getInitialURL: jest.fn(() => Promise.resolve()),
}));
