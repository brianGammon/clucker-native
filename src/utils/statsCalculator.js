import moment from 'moment';
import { sortBy, toPairs, fromPairs } from 'lodash';
import { dateStringAsMoment, nowAsMoment } from './dateHelper';

export default (eggs, range) => {
  let heaviestEgg = null;
  let highestWeight = 0;
  let totalWithWeight = 0;
  let totalWeight = 0;
  let earliestDate = null;
  let rangeCount = 0;
  let daysBackForAvg = 30;
  const now = nowAsMoment();
  const thirtyDaysAgo = now.clone().subtract(daysBackForAvg - 1, 'day');
  const firstOfMonth = range === 'allTime' ? null : moment.utc(`${range}-01`);
  const startofRange = range === 'allTime' ? thirtyDaysAgo : firstOfMonth;
  let eggsPerChicken = {};
  const eggsPerPeriod = {};

  if (range !== 'allTime') {
    if (now.format('YYYY-MM') === range) {
      // in current month, use number of days so far in month
      daysBackForAvg = now.diff(firstOfMonth, 'days') + 1;
    } else {
      // not current month, so use number of days in month
      daysBackForAvg = moment(range).daysInMonth();
    }
  }

  const sortedEggs = sortBy(eggs, 'date');
  if (sortedEggs.length === 0) {
    return null;
  }

  sortedEggs.forEach((egg) => {
    const thisEggDate = moment(egg.date);

    if (!earliestDate) {
      earliestDate = egg.date;
    }

    if (!eggsPerChicken[egg.chickenId]) {
      eggsPerChicken[egg.chickenId] = 0;
    }
    eggsPerChicken[egg.chickenId] += 1;

    // Find heaviest & avg
    if (egg.weight) {
      totalWithWeight += 1;
      // Some weights are stored as strings (TODO)
      totalWeight += +egg.weight;
      if (+egg.weight > highestWeight) {
        heaviestEgg = egg;
        highestWeight = egg.weight;
      }
    }

    if (range === 'allTime' || thisEggDate.isAfter(startofRange)) {
      rangeCount += 1;
      const rollupPeriod = range === 'allTime' ? egg.date.substring(0, 7) : egg.date;

      if (!eggsPerPeriod[rollupPeriod]) {
        eggsPerPeriod[rollupPeriod] = { total: 0, byChicken: {} };
      }

      eggsPerPeriod[rollupPeriod].total += 1;
      eggsPerPeriod[rollupPeriod].byChicken[egg.chickenId] = (eggsPerPeriod[rollupPeriod].byChicken[egg.chickenId] || 0) + 1;
    }
  });

  const earliest = dateStringAsMoment(earliestDate);
  let daysToGoBack = daysBackForAvg;

  if (range === 'allTime' && earliest.isAfter(startofRange)) {
    const daysAfter = earliest.diff(startofRange, 'days');
    daysToGoBack = daysBackForAvg - daysAfter;
  }
  const averageNumber = rangeCount > 0 ? rangeCount / daysToGoBack : 0;

  // sort the eggPerPeriod by most eggs
  const array = toPairs(eggsPerChicken);
  const sortedArray = array.sort((a, b) => b[1] - a[1]);
  const topProducer = sortedArray[0][0];
  eggsPerChicken = fromPairs(sortedArray);

  const stats = {
    total: sortedEggs.length,
    heaviest: heaviestEgg,
    averageWeight: totalWeight / totalWithWeight,
    averageNumber,
    firstEgg: earliestDate,
    mostEggs: topProducer,
    eggsPerChicken,
    eggsPerPeriod,
  };
  return stats;
};
