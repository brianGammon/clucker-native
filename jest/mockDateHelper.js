import moment from 'moment';
import '../src/utils/dateHelper';

jest.mock('../src/utils/dateHelper', () => {
  const module = require.requireActual('../src/utils/dateHelper');
  module.nowAsMoment = () => moment.utc('2018-05-15');
  return module;
});
