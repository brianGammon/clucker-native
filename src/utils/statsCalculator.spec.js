import statsCalculator from './statsCalculator';
import sampleData from '../../jest/test-data/full.2018-04-to-2018-05';
import sampleData2 from '../../jest/test-data/eggs.only-bulk';
import '../../jest/mockDateHelper';

describe('Stats Calculator', () => {
  test('calculate for allTime, full set', () => {
    const stats = statsCalculator(sampleData.eggs.data, 'allTime');
    expect(stats).toMatchSnapshot();
  });

  test('calculate for allTime, only bulk', () => {
    const stats = statsCalculator(sampleData2.eggs.data, 'allTime');
    expect(stats).toMatchSnapshot();
  });
});
