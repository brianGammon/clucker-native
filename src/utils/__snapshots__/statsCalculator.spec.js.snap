// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats Calculator calculate for allTime, full set 1`] = `
Object {
  "averageNumber": 6.166666666666667,
  "averageWeight": 64.09795918367344,
  "eggsPerChicken": Object {
    "chicken1": 36,
    "chicken2": 28,
    "chicken3": 11,
    "chicken4": 31,
    "unknown": 1,
  },
  "eggsPerPeriod": Object {
    "2018-03": Object {
      "byChicken": Object {
        "chicken4": 1,
      },
      "total": 9,
    },
    "2018-04": Object {
      "byChicken": Object {
        "chicken1": 24,
        "chicken2": 20,
        "chicken3": 11,
        "chicken4": 22,
      },
      "hasDamaged": true,
      "total": 218,
    },
    "2018-05": Object {
      "byChicken": Object {
        "chicken1": 12,
        "chicken2": 8,
        "chicken4": 8,
        "unknown": 1,
      },
      "hasDamaged": true,
      "hasNote": true,
      "total": 29,
    },
  },
  "firstEgg": "2018-03-31",
  "heaviest": Object {
    "chickenId": "unknown",
    "date": "2018-05-09",
    "modified": "2018-05-22T23:38:20.234Z",
    "weight": "93.6",
  },
  "mostEggs": "chicken1",
  "total": 256,
}
`;

exports[`Stats Calculator calculate for allTime, only bulk 1`] = `
Object {
  "averageNumber": 2.2,
  "averageWeight": null,
  "eggsPerChicken": Object {},
  "eggsPerPeriod": Object {
    "2018-05": Object {
      "byChicken": Object {},
      "hasNote": true,
      "total": 22,
    },
  },
  "firstEgg": "2018-05-06",
  "heaviest": null,
  "mostEggs": false,
  "total": 22,
}
`;
